
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Introduction / Quick start &#8212; genomvar  documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Variants" href="Variants.html" />
    <link rel="prev" title="genomvar’s documentation" href="index.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Variants.html" title="Variants"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="genomvar’s documentation"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">genomvar  documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="introduction-quick-start">
<h1>Introduction / Quick start<a class="headerlink" href="#introduction-quick-start" title="Permalink to this headline">¶</a></h1>
<p>Package <code class="docutils literal notranslate"><span class="pre">genomvar</span></code> works with genomic variants and implements
set-like operations on them. It supports import from VCF files and
export to NumPy.</p>
<p>For documentation see <a class="reference external" href="https://mikpom.github.io/genomvar/">here</a>.</p>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>Requirements:</p>
<ol class="arabic simple">
<li><p>Python &gt;=3.6</p></li>
<li><p>rbi-tree</p></li>
<li><p>jinja2</p></li>
<li><p>pysam</p></li>
</ol>
<p>To install:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">genomvar</span>
</pre></div>
</div>
</div>
<div class="section" id="sample-usage">
<h2>Sample usage<a class="headerlink" href="#sample-usage" title="Permalink to this headline">¶</a></h2>
<div class="section" id="case-1">
<h3>Case 1<a class="headerlink" href="#case-1" title="Permalink to this headline">¶</a></h3>
<p>Common task in genome variant analysis is: there are two VCF files (for
example obtained from variant caller #1 and caller #2)
and the differences should be investigated.</p>
<p>First we read the VCF files
into genomvar <a class="reference internal" href="Package_contents.html#genomvar.varset.VariantSet" title="genomvar.varset.VariantSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.varset.VariantSet</span></code></a> objects which
hold the variants with underlying data contained in INFO fields:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar.varset</span> <span class="kn">import</span> <span class="n">VariantSet</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vs1</span> <span class="o">=</span> <span class="n">VariantSet</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="s1">&#39;caller1.out.vcf.gz&#39;</span><span class="p">,</span><span class="n">parse_info</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vs2</span> <span class="o">=</span> <span class="n">VariantSet</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="s1">&#39;caller2.out.vcf.gz&#39;</span><span class="p">,</span><span class="n">parse_info</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>To find variants detected by caller #1 but not caller #2 <code class="docutils literal notranslate"><span class="pre">diff</span></code>
method is used. Then differences are exported to <code class="docutils literal notranslate"><span class="pre">numpy</span></code> for futher
analysis:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">diff</span> <span class="o">=</span> <span class="n">vs1</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">vs2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recs</span> <span class="o">=</span> <span class="n">diff</span><span class="o">.</span><span class="n">to_records</span><span class="p">()</span> <span class="c1"># recs is a numpy structured dtype array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recs</span><span class="p">[[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;ref&#39;</span><span class="p">,</span><span class="s1">&#39;alt&#39;</span><span class="p">,</span><span class="s1">&#39;vartype&#39;</span><span class="p">]]</span>
<span class="go">[(&#39;chr1&#39;,  1046755,  1046756, &#39;T&#39;, &#39;G&#39;, &#39;SNP&#39;)</span>
<span class="go"> (&#39;chr1&#39;,  1057987,  1057988, &#39;T&#39;, &#39;C&#39;, &#39;SNP&#39;)</span>
<span class="go">  ...,</span>
<span class="go"> (&#39;chr19&#39;, 56434340, 56434341, &#39;A&#39;, &#39;G&#39;, &#39;SNP&#39;)</span>
<span class="go"> (&#39;chrY&#39;, 56839067, 56839068, &#39;A&#39;, &#39;G&#39;, &#39;SNP&#39;)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recs</span><span class="p">[</span><span class="s1">&#39;INFO&#39;</span><span class="p">][</span><span class="s1">&#39;DP&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="c1"># recs[&#39;INFO&#39;][&#39;DP&#39;] is a numpy ndarray</span>
<span class="go">232.18819746028257</span>
</pre></div>
</div>
</div>
<div class="section" id="case-2">
<h3>Case 2<a class="headerlink" href="#case-2" title="Permalink to this headline">¶</a></h3>
<p>There is a smaller variant file obtained from the data and a bigger one
usually obtained from a database. Variants in the former should be “annotated”
with some data associated with variants in the latter.</p>
<p>This case is different from the previous in that DB file might not
comfortably fit into memory. Class
<a class="reference internal" href="Package_contents.html#genomvar.varset.IndexedVariantFileSet" title="genomvar.varset.IndexedVariantFileSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">IndexedVariantFileSet</span></code></a> can be used for this
purpose:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vs</span> <span class="o">=</span> <span class="n">varset</span><span class="o">.</span><span class="n">VariantSet</span><span class="o">.</span><span class="n">from_vcf</span><span class="p">(</span><span class="s1">&#39;vcf_of_interest.vcf&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dbSNP</span> <span class="o">=</span> <span class="n">varset</span><span class="o">.</span><span class="n">IndexedVariantFileSet</span><span class="p">(</span><span class="s1">&#39;DBSNP.vcf.gz&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">annots</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">vrt</span> <span class="ow">in</span> <span class="n">vs</span><span class="o">.</span><span class="n">iter_vrt</span><span class="p">():</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">m</span> <span class="o">=</span> <span class="n">dbSNP</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">annots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">m</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">attrib</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">m</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">annots</span>
<span class="go">[None, None, &#39;rs540057607&#39;, &#39;rs367710686&#39;, &#39;rs940651103&#39;, ...]</span>
</pre></div>
</div>
<p>Here <a class="reference internal" href="Package_contents.html#genomvar.varset.VariantSet.match" title="genomvar.varset.VariantSet.match"><code class="xref py py-meth docutils literal notranslate"><span class="pre">match()</span></code></a> method is used. It
searches for variants with the same genomic alteration as argument
variant and returns a list of those.  Then VCF <code class="docutils literal notranslate"><span class="pre">ID</span></code> field can be
accessed from those matching variants in <code class="docutils literal notranslate"><span class="pre">attrib['id']</span></code> (dbSNP rs
numbers in this particular case).</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Introduction / Quick start</a><ul>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#sample-usage">Sample usage</a><ul>
<li><a class="reference internal" href="#case-1">Case 1</a></li>
<li><a class="reference internal" href="#case-2">Case 2</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">genomvar’s documentation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Variants.html"
                        title="next chapter">Variants</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/Introduction.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Variants.html" title="Variants"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="genomvar’s documentation"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">genomvar  documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, AcademGene LLC.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.0.0.
    </div>
  </body>
</html>