
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Package contents &#8212; genomvar  documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Comparing two VCF files from command line" href="Command_line_utility.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Command_line_utility.html" title="Comparing two VCF files from command line"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">genomvar  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Package contents</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="package-contents">
<h1>Package contents<a class="headerlink" href="#package-contents" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-genomvar.variant">
<span id="module-variant-py"></span><h2>Module variant.py<a class="headerlink" href="#module-genomvar.variant" title="Permalink to this headline">¶</a></h2>
<p>Module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> contains classes representing genomic
alterations.</p>
<p>The hierarchy of the classes used in the package is the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>                  <span class="n">VariantBase</span>
                  <span class="o">/</span>      <span class="o">|</span>   \
 <span class="n">AmbigIndel</span> <span class="o">&lt;-</span><span class="n">Indel</span>      <span class="n">V</span>   <span class="o">|</span>
   <span class="o">|</span>     <span class="o">|</span>      <span class="o">/</span> \     <span class="n">MNP</span>  <span class="n">V</span>
   <span class="o">|</span> <span class="o">----+---</span> <span class="n">Del</span>  <span class="n">Ins</span>   <span class="o">|</span>  <span class="n">Haplotype</span>
   <span class="o">|</span> <span class="o">|</span>   <span class="o">|</span>          <span class="o">|</span>    <span class="n">V</span>
   <span class="n">V</span> <span class="n">V</span>   <span class="o">|</span>          <span class="n">V</span>   <span class="n">SNP</span>
<span class="n">AmbigDel</span> <span class="o">-----&gt;</span> <span class="n">AmbigIns</span>
</pre></div>
</div>
<p>All variants have <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> attributes defining a range they
act on and can be searched overlap for.</p>
<p>To test whether a variant is instance of some type
<a class="reference internal" href="#genomvar.variant.VariantBase.is_variant_instance" title="genomvar.variant.VariantBase.is_variant_instance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">is_variant_instance()</span></code></a> method can be used.  Variant equality
can be tested using <a class="reference internal" href="#genomvar.variant.VariantBase.edit_equal" title="genomvar.variant.VariantBase.edit_equal"><code class="xref py py-meth docutils literal notranslate"><span class="pre">edit_equal()</span></code></a>.</p>
<p>Objects can be instantiated directly, e.g.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vrt</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">MNP</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span><span class="mi">154678</span><span class="p">,</span><span class="s1">&#39;GT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="go">&lt;MNP chr1:154678-154680 NN/GT&gt;</span>
</pre></div>
</div>
<p>This will create an MNP which substitutes positions 154678 and 154679 on
chromosome 1 for <code class="docutils literal notranslate"><span class="pre">GT</span></code>.</p>
<p>Alternatively variants can be created using <a class="reference internal" href="#genomvar.variant.VariantFactory" title="genomvar.variant.VariantFactory"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantFactory</span></code></a>
objects.  This class can work with VCF-like notation. For example,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fac</span> <span class="o">=</span> <span class="n">VariantFactory</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vrt</span> <span class="o">=</span> <span class="n">fac</span><span class="o">.</span><span class="n">from_edit</span><span class="p">(</span><span class="s1">&#39;chr15&#39;</span><span class="p">,</span><span class="mi">575</span><span class="p">,</span><span class="s1">&#39;TA&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="go">&lt;Del chr15:576-577 A/-&gt;</span>
</pre></div>
</div>
<p>Position is <strong>0-based</strong> so it creates a deletion at position 577
of chromosome 15.</p>
<p>Alternatively, limited subset of HGVS notation is supported (numbering
in HGVS strings is <strong>1-based</strong> following the spec):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vrt</span> <span class="o">=</span> <span class="n">fac</span><span class="o">.</span><span class="n">from_hgvs</span><span class="p">(</span><span class="s1">&#39;chr1:g.15C&gt;A&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="go">&lt;SNP chr1:14 C/A&gt;</span>
</pre></div>
</div>
<p>Variant sets defined in <a class="reference internal" href="Variant_sets.html#module-genomvar.varset" title="genomvar.varset"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.varset</span></code></a> use class <a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomVariant</span></code></a>.
Objects of this class contain genomic alteration (attribute <code class="docutils literal notranslate"><span class="pre">base</span></code>)
and optionally, genotype (attribute <code class="docutils literal notranslate"><span class="pre">GT</span></code>) and other attributes
commonly found in VCF files (attribute <code class="docutils literal notranslate"><span class="pre">attrib</span></code>).  Attribute
<code class="docutils literal notranslate"><span class="pre">base</span></code> is an object of some <code class="docutils literal notranslate"><span class="pre">VariantBase</span></code> subclass (SNPs, Deletions
etc.).</p>
<dl class="py class">
<dt id="genomvar.variant.VariantBase">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">VariantBase</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">ref</span></em>, <em class="sig-param"><span class="n">alt</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.VariantBase" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class for the other genomic variant classes.</p>
<dl class="py method">
<dt id="genomvar.variant.VariantBase.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.VariantBase.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.VariantBase.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.VariantBase.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.VariantBase.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.VariantBase.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.VariantBase.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.VariantBase.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.VariantBase.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.VariantBase.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.MNP">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">MNP</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">alt</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.MNP" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiple-nucleotide polymorphism.  Substitute N nucleotides of the
reference for N other nucleotides.</p>
<p>For instantiation it requires chromosome,position and
alternative sequence. <code class="docutils literal notranslate"><span class="pre">end</span></code> will inferred from <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">alt</span></code>.
<code class="docutils literal notranslate"><span class="pre">ref</span></code> is also optional.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar</span> <span class="kn">import</span> <span class="n">variant</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variant</span><span class="o">.</span><span class="n">MNP</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span><span class="mi">154678</span><span class="p">,</span><span class="s1">&#39;GT&#39;</span><span class="p">)</span>
<span class="go">MNP(&quot;chr1&quot;,154678,&quot;GT&quot;)</span>
</pre></div>
</div>
<dl class="py method">
<dt id="genomvar.variant.MNP.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.MNP.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.MNP.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.MNP.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.MNP.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.MNP.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.MNP.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.MNP.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.MNP.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.MNP.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.SNP">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">SNP</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">alt</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.SNP" title="Permalink to this definition">¶</a></dt>
<dd><p>Single-nucleotide polymorphism.</p>
<p>For instantiation it requires chromosome, position and
alternative sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar</span> <span class="kn">import</span> <span class="n">variant</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variant</span><span class="o">.</span><span class="n">SNP</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span><span class="mi">154678</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">)</span>
<span class="go">SNP(&quot;chr1&quot;,154678,&quot;T&quot;)</span>
</pre></div>
</div>
<dl class="py method">
<dt id="genomvar.variant.SNP.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.SNP.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.SNP.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.SNP.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.SNP.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.SNP.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.SNP.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.SNP.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.SNP.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.SNP.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.Indel">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Indel</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">ref</span></em>, <em class="sig-param"><span class="n">alt</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Indel" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract class to accomodate insertion/deletion subclasses</p>
<dl class="py method">
<dt id="genomvar.variant.Indel.ambig_equal">
<code class="sig-name descname">ambig_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Indel.ambig_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns true if two variants are equal up to indel ambiguity.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Indel.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Indel.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Indel.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Indel.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Indel.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Indel.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Indel.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.Indel.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Indel.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.Indel.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.Ins">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Ins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">alt</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Ins" title="Permalink to this definition">¶</a></dt>
<dd><p>Insertion of nucleotides. For instantiation <code class="docutils literal notranslate"><span class="pre">chrom</span></code>,
<code class="docutils literal notranslate"><span class="pre">start</span></code>  and
inserted sequence (<code class="docutils literal notranslate"><span class="pre">alt</span></code>) are required.</p>
<p>Start and end denote the nucleotide after the inserted sequence,
i.e. <code class="docutils literal notranslate"><span class="pre">start</span></code> is 0-based number of a nucleotide after insertion,
<code class="docutils literal notranslate"><span class="pre">end</span></code> is <code class="docutils literal notranslate"><span class="pre">start+1</span></code> by definition.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar.variant</span> <span class="kn">import</span> <span class="n">Ins</span>
<span class="go"># Insertion of TA before position chr2:100543.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Ins</span><span class="p">(</span><span class="s1">&#39;chr2&#39;</span><span class="p">,</span><span class="mi">100543</span><span class="p">,</span><span class="s1">&#39;TA&#39;</span><span class="p">))</span>
<span class="go">&lt;Ins chr2:100543 -/TA&gt;</span>
</pre></div>
</div>
<dl class="py method">
<dt id="genomvar.variant.Ins.ambig_equal">
<code class="sig-name descname">ambig_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Ins.ambig_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns true if two variants are equal up to indel ambiguity.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Ins.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Ins.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Ins.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Ins.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Ins.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Ins.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Ins.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.Ins.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Ins.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.Ins.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.Del">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Del</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alt</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Del" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletion of nucleotides. For instantiation <code class="docutils literal notranslate"><span class="pre">chrom</span></code>,
<code class="docutils literal notranslate"><span class="pre">start</span></code> (0-based),
and <code class="docutils literal notranslate"><span class="pre">end</span></code> (position <code class="docutils literal notranslate"><span class="pre">end</span></code> is excluded) are required.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar.variant</span> <span class="kn">import</span> <span class="n">Del</span>
<span class="go"># Deletion of 3 nucleotides starting at chr3:7843488 (0-based)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Del</span><span class="p">(</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span><span class="mi">7843488</span><span class="p">,</span><span class="mi">7843488</span><span class="o">+</span><span class="mi">3</span><span class="p">))</span>
<span class="go">&lt;Del chr3:7843488-7843491 NNN/-&gt;</span>
</pre></div>
</div>
<dl class="py method">
<dt id="genomvar.variant.Del.ambig_equal">
<code class="sig-name descname">ambig_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Del.ambig_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns true if two variants are equal up to indel ambiguity.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Del.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Del.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Del.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Del.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Del.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Del.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Del.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.Del.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Del.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.Del.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.AmbigIndel">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">AmbigIndel</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">ref</span></em>, <em class="sig-param"><span class="n">alt</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIndel" title="Permalink to this definition">¶</a></dt>
<dd><p>Class representing indel which position is ambigous.  Ambiguity means the
indel could be applied in any position of some region resulting in the same
alternative sequence.</p>
<dl class="py method">
<dt id="genomvar.variant.AmbigIndel.ambig_equal">
<code class="sig-name descname">ambig_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIndel.ambig_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns true if two variants are equal up to indel ambiguity.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIndel.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIndel.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIndel.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIndel.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIndel.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIndel.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIndel.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.AmbigIndel.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIndel.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.AmbigIndel.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.AmbigIns">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">AmbigIns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">alt</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIns" title="Permalink to this definition">¶</a></dt>
<dd><p>Class representing indel which position is ambigous.  Ambiguity
means the indel could be applied in any position of some region
resulting in the same alternative sequence.</p>
<p>Let the reference file <code class="docutils literal notranslate"><span class="pre">test.fasta</span></code> contain a toy sequence:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seq1</span>
<span class="n">TTTAATA</span>
</pre></div>
</div>
<p>Consider a variant extending  3 <code class="docutils literal notranslate"><span class="pre">T</span></code> s in the beginning by one
more T. It can be done in several places so the corresponding
insertion can be given as and <code class="docutils literal notranslate"><span class="pre">AmbigIns</span></code> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar</span> <span class="kn">import</span> <span class="n">Reference</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar.variant</span> <span class="kn">import</span> <span class="n">VariantFactory</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fac</span> <span class="o">=</span> <span class="n">VariantFactory</span><span class="p">(</span><span class="n">Reference</span><span class="p">(</span><span class="s1">&#39;test.fasta&#39;</span><span class="p">),</span><span class="n">normindel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span> <span class="n">fac</span><span class="o">.</span><span class="n">from_edit</span><span class="p">(</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">,</span><span class="s1">&#39;TT&#39;</span><span class="p">)</span> <span class="p">)</span>
<span class="go">&lt;AmbigIns seq1:0-4(1-2) -/T&gt;</span>
</pre></div>
</div>
<p>Positions 1 and 2 are actual start and end meaning that T
is inserted before nucleotide located 1-2.
Positions 0-4 indicate that start and end can be extended to
these values resulting in the same alteration.</p>
<dl class="py method">
<dt id="genomvar.variant.AmbigIns.seq">
<em class="property">property </em><code class="sig-name descname">seq</code><a class="headerlink" href="#genomvar.variant.AmbigIns.seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Inserted sequence.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIns.ambig_equal">
<code class="sig-name descname">ambig_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIns.ambig_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns true if two variants are equal up to indel ambiguity.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIns.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIns.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIns.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIns.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIns.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigIns.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIns.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.AmbigIns.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigIns.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.AmbigIns.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.AmbigDel">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">AmbigDel</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alt</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigDel" title="Permalink to this definition">¶</a></dt>
<dd><p>Class representing del which position is ambigous.  Ambiguity means
the same number of positions could be deleted in some range
resulting in the same aternative sequence.</p>
<p>Let the reference file <code class="docutils literal notranslate"><span class="pre">test.fasta</span></code> contain a toy sequence:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seq1</span>
<span class="n">TCTTTTTGACTGG</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fac</span> <span class="o">=</span> <span class="n">VariantFactory</span><span class="p">(</span><span class="n">Reference</span><span class="p">(</span><span class="s1">&#39;test.fasta&#39;</span><span class="p">),</span><span class="n">normindel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">&gt;&gt;&gt;</span>
<span class="go">print( fac.from_edit(&#39;seq1&#39;,1,&#39;CTTTTTGAC&#39;,&#39;C&#39;) ) &lt;AmbigDel</span>
<span class="go">seq1:1-11(2-10) TTTTTGAC/-&gt;</span>
</pre></div>
</div>
<p>Deletion of TTTTTGAC starts at 2 and ends on 9th nucleotide
(including 9th resulting in range 2-10).  1-11 denote that start and
end can be extended to these values resulting in the same
alteration.</p>
<dl class="py method">
<dt id="genomvar.variant.AmbigDel.seq">
<em class="property">property </em><code class="sig-name descname">seq</code><a class="headerlink" href="#genomvar.variant.AmbigDel.seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Deleted sequence.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigDel.ambig_equal">
<code class="sig-name descname">ambig_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigDel.ambig_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns true if two variants are equal up to indel ambiguity.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigDel.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigDel.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigDel.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigDel.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigDel.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.AmbigDel.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigDel.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.AmbigDel.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.AmbigDel.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.AmbigDel.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.Mixed">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Mixed</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">alt</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Mixed" title="Permalink to this definition">¶</a></dt>
<dd><p>Combination of Indel and MNP. Usage of this class is discouraged
and exists for compatibility.</p>
<dl class="py method">
<dt id="genomvar.variant.Mixed.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Mixed.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Mixed.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Mixed.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Mixed.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Mixed.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Mixed.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.Mixed.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Mixed.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.Mixed.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.Haplotype">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Haplotype</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">variants</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Haplotype" title="Permalink to this definition">¶</a></dt>
<dd><p>An object representing genome variants on the
same chromosome (or contig).</p>
<p>Can be instantiated from a list of GenomVariant objects using
<a class="reference internal" href="#genomvar.variant.Haplotype.from_variants" title="genomvar.variant.Haplotype.from_variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Haplotype.from_variants()</span></code></a> class method.</p>
<dl class="py method">
<dt id="genomvar.variant.Haplotype.find_vrt">
<code class="sig-name descname">find_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">2147483647</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Haplotype.find_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds variants in specified region of a haplotype.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>int</em>) – Start of search interval. <em>Default: 0</em></p></li>
<li><p><strong>end</strong> (<em>int</em>) – End of search interval. <em>Defaults to MAX_END</em></p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><strong>vrt</strong> (<em>variants</em>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Haplotype.from_variants">
<em class="property">classmethod </em><code class="sig-name descname">from_variants</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">variants</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Haplotype.from_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Create haplotype from a list of variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variants</strong> (<em>list of variants</em>) – Variants to instantiate haplotype from.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>hap</strong> – Haplotype variant object.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.variant.Haplotype" title="genomvar.variant.Haplotype">Haplotype</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Haplotype.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Haplotype.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Haplotype.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Haplotype.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Haplotype.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Haplotype.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Haplotype.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.Haplotype.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Haplotype.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.Haplotype.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.Null">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Null</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">ref</span></em>, <em class="sig-param"><span class="n">alt</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Null" title="Permalink to this definition">¶</a></dt>
<dd><dl class="py method">
<dt id="genomvar.variant.Null.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Null.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Null.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Null.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Null.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Null.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Null.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.Null.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Null.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.Null.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.Asterisk">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Asterisk</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">ref</span></em>, <em class="sig-param"><span class="n">alt</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Asterisk" title="Permalink to this definition">¶</a></dt>
<dd><dl class="py method">
<dt id="genomvar.variant.Asterisk.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Asterisk.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if <code class="docutils literal notranslate"><span class="pre">self</span></code> represents the same alteration as the <code class="docutils literal notranslate"><span class="pre">other</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Asterisk.is_variant_instance">
<code class="sig-name descname">is_variant_instance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cls</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Asterisk.is_variant_instance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if object’s variant class is subclass of
<code class="docutils literal notranslate"><span class="pre">cls</span></code>.  See module <a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentation for
class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Asterisk.is_variant_subclass">
<em class="property">classmethod </em><code class="sig-name descname">is_variant_subclass</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.Asterisk.is_variant_subclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks is variant class is subclass of <code class="docutils literal notranslate"><span class="pre">other</span></code>.  See module
<a class="reference internal" href="#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> documentaion for class hierarchy.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Asterisk.key">
<em class="property">property </em><code class="sig-name descname">key</code><a class="headerlink" href="#genomvar.variant.Asterisk.key" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple representation of variant.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.Asterisk.vtp">
<em class="property">property </em><code class="sig-name descname">vtp</code><a class="headerlink" href="#genomvar.variant.Asterisk.vtp" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant class</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.GenomVariant">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">GenomVariant</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">base</span></em>, <em class="sig-param"><span class="n">attrib</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.GenomVariant" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a variant class holding gemomic alteration
(a VariantBase subclass object)
and extra attributes. Underlying variant is accessible as <code class="docutils literal notranslate"><span class="pre">base</span></code> attribute.
On top it has genotype as <code class="docutils literal notranslate"><span class="pre">GT</span></code> attribute and <code class="docutils literal notranslate"><span class="pre">attrib</span></code> containing
extra attributes parsed from VCF files.  All attributes of
underlying <code class="docutils literal notranslate"><span class="pre">base</span></code> (start, end etc.) are accessible from
GenomVariant object.</p>
<dl class="py method">
<dt id="genomvar.variant.GenomVariant.edit_equal">
<code class="sig-name descname">edit_equal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.GenomVariant.edit_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if GenomVariant holds the same genome alteration
as <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> (<a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><em>GenomVariant</em></a><em> or </em><em>any VariantBase</em>) – variant to compare</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>equality</strong> – True if the same genomic alteration.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.variant.VariantFactory">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">VariantFactory</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.VariantFactory" title="Permalink to this definition">¶</a></dt>
<dd><p>Factory class used to create Variant objects.  Can be instantiated
without arguments or an instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.Reference</span></code> can
be given. Additionnaly if <code class="docutils literal notranslate"><span class="pre">normindel=True</span></code> (reference should be
given) indels will be checked for ambiguity, and if ambigous instance
of <a class="reference internal" href="#genomvar.variant.AmbigIndel" title="genomvar.variant.AmbigIndel"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.AmbigIndel</span></code></a> will be returned.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar</span> <span class="kn">import</span> <span class="n">Reference</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar.variant</span> <span class="kn">import</span> <span class="n">VariantFactory</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference</span> <span class="o">=</span> <span class="n">Reference</span><span class="p">(</span><span class="s1">&#39;test.fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fac</span> <span class="o">=</span> <span class="n">VariantFactory</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span><span class="n">normindel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt id="genomvar.variant.VariantFactory.from_hgvs">
<code class="sig-name descname">from_hgvs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">st</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.VariantFactory.from_hgvs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a variant from a HGVS notation string. Only some of
possible HGVS notations are supported as listed below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fac</span> <span class="o">=</span> <span class="n">VariantFactory</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fac</span><span class="o">.</span><span class="n">from_hgvs</span><span class="p">(</span><span class="s1">&#39;chr1:g.15C&gt;A&#39;</span><span class="p">))</span>
<span class="go">&lt;SNP chr1:14 C/A&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fac</span><span class="o">.</span><span class="n">from_hgvs</span><span class="p">(</span><span class="s1">&#39;chrW:g.19_21del&#39;</span><span class="p">))</span>
<span class="go">&lt;Del chrW:18-21 NNN/-&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fac</span><span class="o">.</span><span class="n">from_hgvs</span><span class="p">(</span><span class="s1">&#39;chr24:g.10_11insCCT&#39;</span><span class="p">))</span>
<span class="go">&lt;Ins chr24:10 -/CCT&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fac</span><span class="o">.</span><span class="n">from_hgvs</span><span class="p">(</span><span class="s1">&#39;chr24:g.10delinsGA&#39;</span><span class="p">))</span>
<span class="go">&lt;Mixed chr24:9-10 -/GA&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fac</span><span class="o">.</span><span class="n">from_hgvs</span><span class="p">(</span><span class="s1">&#39;chr23:g.145_147delinsTGG&#39;</span><span class="p">))</span>
<span class="go">&lt;MNP chr23:144-147 NNN/TGG&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="genomvar.variant.VariantFactory.from_edit">
<code class="sig-name descname">from_edit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">ref</span></em>, <em class="sig-param"><span class="n">alt</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.variant.VariantFactory.from_edit" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes chrom, start, ref and alt (similar to a row in VCF file,
but start is 0-based) and returns a variant object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vrt</span> <span class="o">=</span> <span class="n">fac</span><span class="o">.</span><span class="n">from_edit</span><span class="p">(</span><span class="s1">&#39;chr15&#39;</span><span class="p">,</span><span class="mi">575</span><span class="p">,</span><span class="s1">&#39;TA&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="go">&lt;Del chr15:576-577 A/-&gt;</span>
</pre></div>
</div>
<p>Method attempts to strip ref and alt sequences from left and right
(in this order), for example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vrt</span> <span class="o">=</span> <span class="n">fac</span><span class="o">.</span><span class="n">from_edit</span><span class="p">(</span><span class="s1">&#39;chr15&#39;</span><span class="p">,</span><span class="n">start</span><span class="o">=</span><span class="mi">65</span><span class="p">,</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;CTG&#39;</span><span class="p">,</span><span class="n">alt</span><span class="o">=</span><span class="s1">&#39;CTC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="go">&lt;SNP chr15:67-68 G/C&gt;</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-genomvar.varset">
<span id="module-varset-py"></span><h2>Module varset.py<a class="headerlink" href="#module-genomvar.varset" title="Permalink to this headline">¶</a></h2>
<p>Module <code class="docutils literal notranslate"><span class="pre">varset</span></code> defines variant set classes which are containers of variants
supporting set-like operations, searching variants, export to NumPy…</p>
<p>There are two classes depending on whether all the variants are loaded
into memory or not:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="#genomvar.varset.VariantSet" title="genomvar.varset.VariantSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantSet</span></code></a> in-memory</p></li>
<li><p><a class="reference internal" href="#genomvar.varset.VariantSetFromFile" title="genomvar.varset.VariantSetFromFile"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantSetFromFile</span></code></a> can read VCF files
on demand, needs index for random access</p></li>
</ul>
</div></blockquote>
<dl class="py class">
<dt id="genomvar.varset.VariantSet">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.varset.</code><code class="sig-name descname">VariantSet</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">variants</span></em>, <em class="sig-param"><span class="n">attrib</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">vcf_notation</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">info</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sampdata</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dtype</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Immutable class for storing variants in-memory.</p>
<p>Supports random access by variant location.
Can be instantiated from a VCF file (see
<a class="reference internal" href="#genomvar.varset.VariantSet.from_vcf" title="genomvar.varset.VariantSet.from_vcf"><code class="xref py py-meth docutils literal notranslate"><span class="pre">VariantSet.from_vcf()</span></code></a>) or from iterable
containing variants (see <a class="reference internal" href="#genomvar.varset.VariantSet.from_variants" title="genomvar.varset.VariantSet.from_variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">VariantSet.from_variants()</span></code></a>).</p>
<dl class="py attribute">
<dt id="genomvar.varset.VariantSet.reference">
<code class="sig-name descname">reference</code><a class="headerlink" href="#genomvar.varset.VariantSet.reference" title="Permalink to this definition">¶</a></dt>
<dd><p>Genome reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Reference</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="genomvar.varset.VariantSet.ctg_len">
<code class="sig-name descname">ctg_len</code><a class="headerlink" href="#genomvar.varset.VariantSet.ctg_len" title="Permalink to this definition">¶</a></dt>
<dd><p>dict mapping Chromosome name to corresponding lengths.</p>
<p>This keeps track of the rightmost position of variants in the set
if reference was not provided. If reference was given <code class="docutils literal notranslate"><span class="pre">ctg_len</span></code>
is taken from length of chromosomes in the reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.sample">
<code class="sig-name descname">sample</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">size</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a random sample of variants.</p>
<p>Haplotypes (if present) maybe returned equally likely as non-haplotype
variants and their subvariants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>size</strong> (<em>int</em>) – size of the sample</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>marray</strong> – List of variants.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.iter_vrt">
<code class="sig-name descname">iter_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">expand</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.iter_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over all variants in the set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>expand</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> variants in encountered haplotypes are yielded.
<em>Default: False</em></p>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.GenomVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of variant set</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.nof_unit_vrt">
<code class="sig-name descname">nof_unit_vrt</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.nof_unit_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns number of simple genomic variations in a variant set.</p>
<p>SNPs and indels count as 1. MNPs as number of affected nucleotides.
Mixed variants (if present) are count also as 1.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.drop_duplicates">
<code class="sig-name descname">drop_duplicates</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">return_dropped</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.drop_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove non-unique variants.</p>
<p>Returns a variant set where no pair of variants
is edit-equal. Haplotypes are left as is.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>return_dropped</strong> (<em>bool</em>) – Whether to return a list of dropped variants.  <em>Default: False</em></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>vs</strong> (<em>VariantSet</em>) – Variant set with unique variants.</p></li>
<li><p><strong>dropped</strong> (<em>list of variants, optional</em>) – if <code class="docutils literal notranslate"><span class="pre">return_dropped</span></code> is True dropped variants are also returned.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.from_vcf">
<em class="property">classmethod </em><code class="sig-name descname">from_vcf</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">vcf</span></em>, <em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">duplicates</span><span class="o">=</span><span class="default_value">'ignore'</span></em>, <em class="sig-param"><span class="n">parse_null</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.from_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse VCF variant file and return VariantSet object</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vcf</strong> (<em>str</em>) – VCF file to read data from.</p></li>
<li><p><strong>reference</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.Reference</span></code>, optional) – Reference genome. <em>Default: None</em></p></li>
<li><p><strong>parse_info</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> INFO fields will be parsed.
<em>Default: False</em>.</p></li>
<li><p><strong>parse_samples</strong> (<em>bool</em><em>, </em><em>str</em><em> or </em><em>list of str</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> all sample data will be parsed.  If string, sample
with this name will be parsed.  If list of strings only these
samples will be parsed. <em>Default: False</em></p></li>
<li><p><strong>normindel</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> indels will be normalized. <code class="docutils literal notranslate"><span class="pre">reference</span></code> should also
be provided.  <em>Default: False</em></p></li>
<li><p><strong>duplicates</strong> (<em>{'ignore'</em><em>,</em><em>'keepfirst'</em><em>,</em><em>'raise'}</em>) – How to treat duplicate variants. If <code class="docutils literal notranslate"><span class="pre">keepfirst</span></code> only the first
is taken. If <code class="docutils literal notranslate"><span class="pre">raise</span></code> error is raised. <em>Default: ``ignore`` and
do not check for duplicates</em>.</p></li>
<li><p><strong>parse_null</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> null variants will also be parsed (experimental).
<em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.varset.VariantSet" title="genomvar.varset.VariantSet">VariantSet</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.from_variants">
<em class="property">classmethod </em><code class="sig-name descname">from_variants</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">variants</span></em>, <em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.from_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Instantiate VariantSet from iterable of variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variants</strong> (<em>iterable</em>) – Variants to add.</p></li>
<li><p><strong>reference</strong> (<em>Reference</em><em>, </em><em>optional</em>) – Genome reference <em>Default: None</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.varset.VariantSet" title="genomvar.varset.VariantSet">VariantSet</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.to_records">
<code class="sig-name descname">to_records</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nested</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.to_records" title="Permalink to this definition">¶</a></dt>
<dd><p>Export <code class="docutils literal notranslate"><span class="pre">self</span></code> to NumPy ndarray.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>nested</strong> (<em>bool</em><em>, </em><em>optional</em>) – Matters only if <code class="docutils literal notranslate"><span class="pre">self</span></code> was instatiated with INFO or SAMPDATA.
When <code class="docutils literal notranslate"><span class="pre">True</span></code> dtype of the return will be nested. If <code class="docutils literal notranslate"><span class="pre">False</span></code>
no recursive fields will be present and “<a href="#id1"><span class="problematic" id="id2">info_</span></a>” or “SAMP_%SAMP%” will
be prepended to corresponding fields <em>Default: True</em></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>NumPy ndarray with structured dtype</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.chroms">
<em class="property">property </em><code class="sig-name descname">chroms</code><a class="headerlink" href="#genomvar.varset.VariantSet.chroms" title="Permalink to this definition">¶</a></dt>
<dd><p>Chromosome set</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.sort_chroms">
<code class="sig-name descname">sort_chroms</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">key</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.sort_chroms" title="Permalink to this definition">¶</a></dt>
<dd><p>Sorts chromosomes in the set</p>
<p>This can influence order in which methods return variants and
variant set comparison.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>key</strong> (<em>callable</em><em>, </em><em>optional</em>) – key to use for chromosome sorting. If not given will sort
lexicographically. <em>Default: None</em></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.diff">
<code class="sig-name descname">diff</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">match_partial</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.diff" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new variant set object which has variants
present in <code class="docutils literal notranslate"><span class="pre">self</span></code> but absent in the <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>match_partial</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> common positions won’t be searched in non-identical
MNPs. <em>Default: True</em></p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> ambigous indels will be treated as regular indels.
<em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>diff</strong> – different variants</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.varset.VariantSet" title="genomvar.varset.VariantSet">VariantSet</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Here is an example of <code class="docutils literal notranslate"><span class="pre">diff</span></code> operation on two variant sets s1 and s2.
Positions identical to REF are empty:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">REF</span>         <span class="n">GATTGGTAC</span>
<span class="o">---------------------</span>
<span class="n">s1</span>          <span class="n">C</span>  <span class="n">CCC</span>  <span class="n">T</span>
               <span class="n">CCC</span>  <span class="n">G</span>
<span class="o">---------------------</span>
<span class="n">s2</span>                  <span class="n">T</span>
              <span class="n">AC</span>    <span class="n">T</span>
<span class="o">=====================</span>
<span class="n">s1</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">s2</span><span class="p">)</span> <span class="n">C</span>   <span class="n">CC</span>
                <span class="n">CC</span>  <span class="n">G</span>
<span class="o">---------------------</span>
<span class="n">s2</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">s1</span><span class="p">)</span>   <span class="n">A</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.comm">
<code class="sig-name descname">comm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">match_partial</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.comm" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a new variant set object which has variants
present both <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>match_partial</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> common positions won’t be searched in non-identical
MNPs. <em>Default: True</em></p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> ambigous indels will be treated as regular indels.
<em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>comm</strong> – Common variants.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.varset.VariantSet" title="genomvar.varset.VariantSet">VariantSet</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Here is an example of <code class="docutils literal notranslate"><span class="pre">diff</span></code> operation on two variant sets s1 and s2.
Positions identical to REF are empty:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">REF</span>         <span class="n">GATTGGTAC</span>
<span class="o">---------------------</span>
<span class="n">s1</span>          <span class="n">C</span>  <span class="n">CCC</span>  <span class="n">T</span>
               <span class="n">CCC</span>  <span class="n">G</span>
<span class="o">---------------------</span>
<span class="n">s2</span>                  <span class="n">T</span>
              <span class="n">AC</span>    <span class="n">T</span>
<span class="o">=====================</span>
<span class="n">s1</span><span class="o">.</span><span class="n">comm</span><span class="p">(</span><span class="n">s2</span><span class="p">)</span>    <span class="n">C</span>    <span class="n">T</span>
               <span class="n">C</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.comm_vrt">
<code class="sig-name descname">comm_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">match_partial</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.comm_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate variants common between <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
<p>This creates a lazily evaluated object of class <a class="reference internal" href="#genomvar.varset.CmpSet" title="genomvar.varset.CmpSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">CmpSet</span></code></a>. This
object can be further inspected with <code class="docutils literal notranslate"><span class="pre">.iter_vrt()</span></code> or <code class="docutils literal notranslate"><span class="pre">.region()</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<em>variant set</em>) – Variant set to compare against <code class="docutils literal notranslate"><span class="pre">self</span></code></p></li>
<li><p><strong>match_partial</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> common positions won’t be searched in non-identical
MNPs. <em>Default: True</em></p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> ambigous indels will be treated as regular indels.
<em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Transient comparison set</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.varset.CmpSet" title="genomvar.varset.CmpSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">CmpSet</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.diff_vrt">
<code class="sig-name descname">diff_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">match_partial</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.diff_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate variants different between <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
<p>This creates a lazily evaluated object of class <a class="reference internal" href="#genomvar.varset.CmpSet" title="genomvar.varset.CmpSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">CmpSet</span></code></a>. This
object can be further inspected with <code class="docutils literal notranslate"><span class="pre">.iter_vrt()</span></code> or <code class="docutils literal notranslate"><span class="pre">.region()</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<em>variant set</em>) – Variant set to compare against <code class="docutils literal notranslate"><span class="pre">self</span></code></p></li>
<li><p><strong>match_partial</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> common positions won’t be searched in non-identical
MNPs. <em>Default: True</em></p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> ambigous indels will be treated as regular indels.
<em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Transient comparison set</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.varset.CmpSet" title="genomvar.varset.CmpSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">CmpSet</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.find_vrt">
<code class="sig-name descname">find_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">2147483647</span></em>, <em class="sig-param"><span class="n">rgn</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">expand</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.find_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds variants in specified region.</p>
<p>If not <code class="docutils literal notranslate"><span class="pre">chrom</span></code> nor <code class="docutils literal notranslate"><span class="pre">rgn</span></code> parameters are given it
traverses all variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<em>str</em><em>, </em><em>optional</em>) – Chromosome to search variants on. If not given, all chroms
are traversed and <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> parameters ignored.</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start of search region (0-based). <em>Default: 0</em></p></li>
<li><p><strong>end</strong> (<em>int</em><em>, </em><em>optional</em>) – End of search region (position <code class="docutils literal notranslate"><span class="pre">end</span></code> is excluded). <em>Default:
max chrom length</em></p></li>
<li><p><strong>rgn</strong> (<em>str</em><em>, </em><em>optional</em>) – If given chrom, start and end parameters are
ignored.  A string specifying search region, e.g.
<code class="docutils literal notranslate"><span class="pre">chr15:365,008-365,958</span></code>.</p></li>
<li><p><strong>expand</strong> (<em>bool</em><em>, </em><em>optional</em>) – Indicates whether to traverse haplotypes. <em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>List of <a class="reference internal" href="#genomvar.variant.VariantBase" title="genomvar.variant.VariantBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.VariantBase</span></code></a> subclass objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.get_factory">
<code class="sig-name descname">get_factory</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.get_factory" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a VariantFactory object. It will inherit the same reference
as was used for instantiation of <code class="docutils literal notranslate"><span class="pre">self</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.iter_vrt_by_chrom">
<code class="sig-name descname">iter_vrt_by_chrom</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwds</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.iter_vrt_by_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates variants grouped by chromosome.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.match">
<code class="sig-name descname">match</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">vrt</span></em>, <em class="sig-param"><span class="n">match_partial</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.match" title="Permalink to this definition">¶</a></dt>
<dd><p>Find variants mathing vrt.</p>
<p>The method checks whether variant <code class="docutils literal notranslate"><span class="pre">vrt</span></code>
is present in the set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vrt</strong> (<em>VariantBase-like</em>) – Variant to search match for.</p></li>
<li><p><strong>match_partial</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> common positions won’t be searched in non-identical
MNPs (relevant if <code class="docutils literal notranslate"><span class="pre">vrt</span></code> is MNP or SNP or haplotype, containing
these). <em>Default: True</em></p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> ambigous indels will be treated as regular indels
(relevant if <code class="docutils literal notranslate"><span class="pre">vrt</span></code> is Ambigindel or haplotype, containing ones).
<em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict is returned only of <code class="docutils literal notranslate"><span class="pre">vrt</span></code> is a Haplotype</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list or dict with matching variants</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.ovlp">
<code class="sig-name descname">ovlp</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">vrt0</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.ovlp" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns variants altering the same positions as <code class="docutils literal notranslate"><span class="pre">vrt0</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vrt0</strong> (<em>VariantBase-like</em>) – Variant to search overlap with.</p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em>) – If False ambigous indels treated as regular indels. <em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of variants</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSet.to_vcf">
<code class="sig-name descname">to_vcf</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">out</span></em>, <em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">info_spec</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">format_spec</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSet.to_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a minimal VCF with variants from the set.</p>
<p>INFO and SAMPLE data from source data is not preserved.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>out</strong> (<em>handle</em><em> or </em><em>str</em>) – If string then it’s path to file, otherwise a handle
to write variants.</p></li>
<li><p><strong>reference</strong> (<em>Reference</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – If string then it’s path to reference FASTA. Otherwise
object of <code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.Reference</span></code> is expected. Not
necessary if self was instantiated with a reference.
If given, this argument takes precedence over the ref
used at instantiation.</p></li>
<li><p><strong>info_spec</strong> (<em>list of tuples containing specification of</em>) – an INFO field in Order per VCF spec, i.e. <cite>NAME</cite>,
<cite>NUMBER</cite>,  <cite>TYPE</cite>,  <cite>DESCRIPTION</cite>,  <cite>SOURCE</cite>,  <cite>VERSION</cite>.
Only the first three fiels (name, number and type)
are required. If variant set was instantiated from VCF
file spec obtained from it will be used. Providing
empty sequence will omit INFO fields in output.</p></li>
<li><p><strong>format_spec</strong> (<em>spec of FORMAT fields</em><em>, </em><em>similar to info_spec</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s1">&#39;wt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">vs</span><span class="o">.</span><span class="n">to_vcf</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">info_spec</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;DP4&#39;</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="s1">&#39;Integer&#39;</span><span class="p">),</span>
<span class="gp">... </span>                             <span class="p">(</span><span class="s1">&#39;NSV&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;Integer&#39;</span><span class="p">)])</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.varset.CmpSet">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.varset.</code><code class="sig-name descname">CmpSet</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">left</span></em>, <em class="sig-param"><span class="n">right</span></em>, <em class="sig-param"><span class="n">op</span></em>, <em class="sig-param"><span class="n">match_partial</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.CmpSet" title="Permalink to this definition">¶</a></dt>
<dd><p>Class <code class="docutils literal notranslate"><span class="pre">CmpSet</span></code> is a lazy implementation of variant set comparison which is
returned by methods <code class="docutils literal notranslate"><span class="pre">diff_vrt</span></code> and <code class="docutils literal notranslate"><span class="pre">comm_vrt</span></code> of any variant set class
instance. Useful for comparisons including large VCF files while keeping
memory profile low.</p>
<dl class="py method">
<dt id="genomvar.varset.CmpSet.region">
<code class="sig-name descname">region</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">2147483647</span></em>, <em class="sig-param"><span class="n">rgn</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.CmpSet.region" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates variants corresponding to comparison within a region.</p>
<p>Region can be specified using chrom,start,end or rgn with notation like
<code class="docutils literal notranslate"><span class="pre">chr15:1000-2000</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<em>str</em>) – Chromosome</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start of region. <em>Default: 0</em></p></li>
<li><p><strong>end</strong> (<em>int</em>) – End of region. <em>Default: max chrom length</em></p></li>
<li><p><strong>rgn</strong> (<em>str</em>) – Region to yield variants from</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.GenomVariant</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.CmpSet.iter_vrt">
<code class="sig-name descname">iter_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">callback</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.CmpSet.iter_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates variants corresponding to comparison.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>callback</strong> (<em>callable</em>) – A function to be called on a match in right variants set
for every variant of left variant set. Result will be stored
in <code class="docutils literal notranslate"><span class="pre">vrt.attrib['cmp']</span></code> of yielded variant</p>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><strong>variants</strong> (<a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.GenomVariant</span></code></a> or             tuple (int, <a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.GenomVariant</span></code></a>)) – if <code class="docutils literal notranslate"><span class="pre">action</span></code> is <code class="docutils literal notranslate"><span class="pre">diff</span></code> or <code class="docutils literal notranslate"><span class="pre">comm</span></code> function yields instances of
<a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.GenomVariant</span></code></a>. If <code class="docutils literal notranslate"><span class="pre">action</span></code> is <code class="docutils literal notranslate"><span class="pre">all</span></code>
function yields tuples where <code class="docutils literal notranslate"><span class="pre">int</span></code> is 0 if variant is present
in first set; 1 if in second; 2 if in both.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.varset.VariantSetFromFile">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.varset.</code><code class="sig-name descname">VariantSetFromFile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file</span></em>, <em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">index</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant set representing variants contained in underlying VCF file specified
on initialization.</p>
<dl class="py attribute">
<dt id="genomvar.varset.VariantSetFromFile.file">
<code class="sig-name descname">file</code><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.file" title="Permalink to this definition">¶</a></dt>
<dd><p>Variant file storing variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.get_factory">
<code class="sig-name descname">get_factory</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.get_factory" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a VariantFactory object. It will inherit the same reference
as was used for instantiation of <code class="docutils literal notranslate"><span class="pre">self</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.iter_vrt_by_chrom">
<code class="sig-name descname">iter_vrt_by_chrom</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.iter_vrt_by_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates variants grouped by chromosome.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.comm_vrt">
<code class="sig-name descname">comm_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">match_partial</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.comm_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate variants common between <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
<p>This creates a lazily evaluated object of class <a class="reference internal" href="#genomvar.varset.CmpSet" title="genomvar.varset.CmpSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">CmpSet</span></code></a>. This
object can be further inspected with <code class="docutils literal notranslate"><span class="pre">.iter_vrt()</span></code> or <code class="docutils literal notranslate"><span class="pre">.region()</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<em>variant set</em>) – Variant set to compare against <code class="docutils literal notranslate"><span class="pre">self</span></code></p></li>
<li><p><strong>match_partial</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> common positions won’t be searched in non-identical
MNPs. <em>Default: True</em></p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> ambigous indels will be treated as regular indels.
<em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Transient comparison set</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.varset.CmpSet" title="genomvar.varset.CmpSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">CmpSet</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.diff_vrt">
<code class="sig-name descname">diff_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">match_partial</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.diff_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate variants different between <code class="docutils literal notranslate"><span class="pre">self</span></code> and <code class="docutils literal notranslate"><span class="pre">other</span></code>.</p>
<p>This creates a lazily evaluated object of class <a class="reference internal" href="#genomvar.varset.CmpSet" title="genomvar.varset.CmpSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">CmpSet</span></code></a>. This
object can be further inspected with <code class="docutils literal notranslate"><span class="pre">.iter_vrt()</span></code> or <code class="docutils literal notranslate"><span class="pre">.region()</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<em>variant set</em>) – Variant set to compare against <code class="docutils literal notranslate"><span class="pre">self</span></code></p></li>
<li><p><strong>match_partial</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> common positions won’t be searched in non-identical
MNPs. <em>Default: True</em></p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> ambigous indels will be treated as regular indels.
<em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Transient comparison set</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.varset.CmpSet" title="genomvar.varset.CmpSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">CmpSet</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.find_vrt">
<code class="sig-name descname">find_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">2147483647</span></em>, <em class="sig-param"><span class="n">rgn</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">expand</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.find_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds variants in specified region.</p>
<p>If not <code class="docutils literal notranslate"><span class="pre">chrom</span></code> nor <code class="docutils literal notranslate"><span class="pre">rgn</span></code> parameters are given it
traverses all variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<em>str</em><em>, </em><em>optional</em>) – Chromosome to search variants on. If not given, all chroms
are traversed and <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> parameters ignored.</p></li>
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start of search region (0-based). <em>Default: 0</em></p></li>
<li><p><strong>end</strong> (<em>int</em><em>, </em><em>optional</em>) – End of search region (position <code class="docutils literal notranslate"><span class="pre">end</span></code> is excluded). <em>Default:
max chrom length</em></p></li>
<li><p><strong>rgn</strong> (<em>str</em><em>, </em><em>optional</em>) – If given chrom, start and end parameters are
ignored.  A string specifying search region, e.g.
<code class="docutils literal notranslate"><span class="pre">chr15:365,008-365,958</span></code>.</p></li>
<li><p><strong>expand</strong> (<em>bool</em><em>, </em><em>optional</em>) – Indicates whether to traverse haplotypes. <em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>List of <a class="reference internal" href="#genomvar.variant.VariantBase" title="genomvar.variant.VariantBase"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.VariantBase</span></code></a> subclass objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.iter_vrt">
<code class="sig-name descname">iter_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">expand</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.iter_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over all variants in underlying VCF.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>expand</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> variants in encountered haplotypes are yielded.
<em>Default: False</em></p>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><strong>variant</strong> (<a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.GenomVariant</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.match">
<code class="sig-name descname">match</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">vrt</span></em>, <em class="sig-param"><span class="n">match_partial</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.match" title="Permalink to this definition">¶</a></dt>
<dd><p>Find variants mathing vrt.</p>
<p>The method checks whether variant <code class="docutils literal notranslate"><span class="pre">vrt</span></code>
is present in the set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vrt</strong> (<em>VariantBase-like</em>) – Variant to search match for.</p></li>
<li><p><strong>match_partial</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> common positions won’t be searched in non-identical
MNPs (relevant if <code class="docutils literal notranslate"><span class="pre">vrt</span></code> is MNP or SNP or haplotype, containing
these). <em>Default: True</em></p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">False</span></code> ambigous indels will be treated as regular indels
(relevant if <code class="docutils literal notranslate"><span class="pre">vrt</span></code> is Ambigindel or haplotype, containing ones).
<em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict is returned only of <code class="docutils literal notranslate"><span class="pre">vrt</span></code> is a Haplotype</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list or dict with matching variants</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.ovlp">
<code class="sig-name descname">ovlp</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">vrt0</span></em>, <em class="sig-param"><span class="n">match_ambig</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.ovlp" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns variants altering the same positions as <code class="docutils literal notranslate"><span class="pre">vrt0</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vrt0</strong> (<em>VariantBase-like</em>) – Variant to search overlap with.</p></li>
<li><p><strong>match_ambig</strong> (<em>bool</em>) – If False ambigous indels treated as regular indels. <em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list of variants</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.to_vcf">
<code class="sig-name descname">to_vcf</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">out</span></em>, <em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">info_spec</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">format_spec</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.to_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a minimal VCF with variants from the set.</p>
<p>INFO and SAMPLE data from source data is not preserved.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>out</strong> (<em>handle</em><em> or </em><em>str</em>) – If string then it’s path to file, otherwise a handle
to write variants.</p></li>
<li><p><strong>reference</strong> (<em>Reference</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – If string then it’s path to reference FASTA. Otherwise
object of <code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.Reference</span></code> is expected. Not
necessary if self was instantiated with a reference.
If given, this argument takes precedence over the ref
used at instantiation.</p></li>
<li><p><strong>info_spec</strong> (<em>list of tuples containing specification of</em>) – an INFO field in Order per VCF spec, i.e. <cite>NAME</cite>,
<cite>NUMBER</cite>,  <cite>TYPE</cite>,  <cite>DESCRIPTION</cite>,  <cite>SOURCE</cite>,  <cite>VERSION</cite>.
Only the first three fiels (name, number and type)
are required. If variant set was instantiated from VCF
file spec obtained from it will be used. Providing
empty sequence will omit INFO fields in output.</p></li>
<li><p><strong>format_spec</strong> (<em>spec of FORMAT fields</em><em>, </em><em>similar to info_spec</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s1">&#39;wt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">vs</span><span class="o">.</span><span class="n">to_vcf</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="n">info_spec</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;DP4&#39;</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="s1">&#39;Integer&#39;</span><span class="p">),</span>
<span class="gp">... </span>                             <span class="p">(</span><span class="s1">&#39;NSV&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;Integer&#39;</span><span class="p">)])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="genomvar.varset.VariantSetFromFile.chroms">
<em class="property">property </em><code class="sig-name descname">chroms</code><a class="headerlink" href="#genomvar.varset.VariantSetFromFile.chroms" title="Permalink to this definition">¶</a></dt>
<dd><p>Chromosome set</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-genomvar.vcf">
<span id="module-vcf-py"></span><h2>Module vcf.py<a class="headerlink" href="#module-genomvar.vcf" title="Permalink to this headline">¶</a></h2>
<p>Module contains classes needed for VCF parsing.</p>
<p>Main class is <a class="reference internal" href="#genomvar.vcf.VCFReader" title="genomvar.vcf.VCFReader"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFReader</span></code></a> which is instantiated from VCF file.
VCF can be gzipped. Bgzipping and tabix-derived indexing is also
supported for random coordinate-based access.</p>
<p><a class="reference internal" href="#genomvar.vcf.VCFReader" title="genomvar.vcf.VCFReader"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFReader</span></code></a> class can iterate over rows which are tuple-like object
containing VCF field strings as attributes without conversion (except
POSition which is converted to int).</p>
<p>Alternatively iteration of variants is supported. It yields
<a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.GenomVariant</span></code></a> objects.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reader</span> <span class="o">=</span> <span class="n">VCFReader</span><span class="p">(</span><span class="s1">&#39;example.vcf.gz&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vrt</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">reader</span><span class="o">.</span><span class="n">iter_vrt</span><span class="p">(</span><span class="n">parse_info</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parse_samples</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="go">&lt;GenomVariant: Del chr24:23-24 G/-&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="o">.</span><span class="n">attrib</span><span class="p">[</span><span class="s1">&#39;info&#39;</span><span class="p">])</span>
<span class="go">{&#39;NSV&#39;: 2, &#39;AF&#39;: 0.5, &#39;DP4&#39;: (11, 22, 33, 44), &#39;ECNT&#39;: 1,    &#39;pl&#39;: 3, &#39;mt&#39;: &#39;SUBSTITUTE&#39;, &#39;RECN&#39;: 18, &#39;STR&#39;: None}</span>
</pre></div>
</div>
<p>If <code class="docutils literal notranslate"><span class="pre">parse_info</span></code> and <code class="docutils literal notranslate"><span class="pre">parse_samples</span></code> parameters are <code class="docutils literal notranslate"><span class="pre">True</span></code> Then
INFO and SAMPLEs fields contained in VCF are parsed and split
corresponding to an allele captured by variant object. For performance
reasons these parameters are set to False and these fields are not
parsed.</p>
<p>.bcf format is supported, use <a class="reference internal" href="#genomvar.vcf.BCFReader" title="genomvar.vcf.BCFReader"><code class="xref py py-class docutils literal notranslate"><span class="pre">BCFReader</span></code></a>.</p>
<dl class="py class">
<dt id="genomvar.vcf.VCFWriter">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.vcf.</code><code class="sig-name descname">VCFWriter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">info_spec</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">format_spec</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for writing variant to VCF format.</p>
<dl class="py method">
<dt id="genomvar.vcf.VCFWriter.get_row">
<code class="sig-name descname">get_row</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">vrt</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwds</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFWriter.get_row" title="Permalink to this definition">¶</a></dt>
<dd><p>Formats a variant to a <a class="reference internal" href="#genomvar.vcf_utils.VCFRow" title="genomvar.vcf_utils.VCFRow"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.vcf_utils.VCFRow</span></code></a> instance.</p>
<p>For indels writer with reference
might be needeed to costruct correct REF field.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vrt</strong> (<em>Variant instance</em>) – variant to get row for.</p></li>
<li><p><strong>kwds</strong> (<em>VCF fields #FIXME remove kwds here</em>) – <p>optional. If given, these and only these parameters are
used to populate corresponding VCF fields: <code class="docutils literal notranslate"><span class="pre">id</span></code>,
<code class="docutils literal notranslate"><span class="pre">qual</span></code>, <code class="docutils literal notranslate"><span class="pre">filter</span></code>, <code class="docutils literal notranslate"><span class="pre">info</span></code>, <code class="docutils literal notranslate"><span class="pre">sampdata</span></code>.
These parameters are taken as is and converted to string
before returning a VCFRow.
If the keyword is not given corresponding field will be
populated from <code class="docutils literal notranslate"><span class="pre">attrib</span></code> attribute if possible.</p>
<p>Any other keyword arguments are ignored.</p>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>row</strong> – <a class="reference internal" href="#genomvar.vcf_utils.VCFRow" title="genomvar.vcf_utils.VCFRow"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.vcf_utils.VCFRow</span></code></a> object instance</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#genomvar.vcf_utils.VCFRow" title="genomvar.vcf_utils.VCFRow">VCFRow</a></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">factory</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">VariantFactory</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">writer</span> <span class="o">=</span> <span class="n">VCFWriter</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">v1</span> <span class="o">=</span> <span class="n">factory</span><span class="o">.</span><span class="n">from_edit</span><span class="p">(</span><span class="s1">&#39;chr24&#39;</span><span class="p">,</span> <span class="mi">2093</span><span class="p">,</span> <span class="s1">&#39;TGG&#39;</span><span class="p">,</span> <span class="s1">&#39;CCC&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span> <span class="o">=</span> <span class="n">writer</span><span class="o">.</span><span class="n">get_row</span><span class="p">(</span><span class="n">v1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">row</span>
<span class="go">&lt;VCFRow chr24:2094 TGG-&gt;CCC&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
<span class="go">chr24   2094    .       TGG     CCC     .       .       .</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">writer</span><span class="o">.</span><span class="n">get_row</span><span class="p">(</span><span class="n">v1</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">info</span><span class="o">=</span><span class="s1">&#39;DP=10&#39;</span><span class="p">))</span>
<span class="go">chr24   2094    123     TGG     CCC     .       .       DP=10</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.vcf.VCFReader">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.vcf.</code><code class="sig-name descname">VCFReader</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">vcf</span></em>, <em class="sig-param"><span class="n">index</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to read VCF files.</p>
<dl class="py method">
<dt id="genomvar.vcf.VCFReader.get_factory">
<code class="sig-name descname">get_factory</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader.get_factory" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a factory based on normindel parameter.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.VCFReader.find_rows">
<code class="sig-name descname">find_rows</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">rgn</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader.find_rows" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields rows of variant file</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.VCFReader.iter_rows">
<code class="sig-name descname">iter_rows</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader.iter_rows" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields rows of variant file</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.VCFReader.iter_rows_by_chrom">
<code class="sig-name descname">iter_rows_by_chrom</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader.iter_rows_by_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields rows grouped by chromosome</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.VCFReader.get_records">
<code class="sig-name descname">get_records</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader.get_records" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns parsed variant data as a dict of NumPy arrays with structured
dtype.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.VCFReader.iter_vrt">
<code class="sig-name descname">iter_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader.iter_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields variant objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parse_info</strong> (<em>bool</em>) – Whether INFO fields should be parsed.  <em>Default: False</em></p></li>
<li><p><strong>parse_samples</strong> (<em>bool</em>) – Whether SAMPLEs dat should be parsed.  <em>Default: False</em></p></li>
<li><p><strong>check_order</strong> (<em>bool</em>) – If True will raise exception on unsorted VCF rows. <em>Default: False</em></p></li>
<li><p><strong>normindel</strong> (<em>bool</em>) – If True insertions and deletions will be left normalized.
Requires a reference on <a class="reference internal" href="#genomvar.vcf.VCFReader" title="genomvar.vcf.VCFReader"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFReader</span></code></a> instantiation.</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><strong>vrt</strong> (<a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.GenomVariant</span></code></a>) – Variant object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.VCFReader.iter_vrt_by_chrom">
<code class="sig-name descname">iter_vrt_by_chrom</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader.iter_vrt_by_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates variants grouped by chromosome</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parse_info</strong> (<em>bool</em>) – Incates whether INFO fields should be parsed.  <em>Default: False</em></p></li>
<li><p><strong>parse_samples</strong> (<em>bool</em>) – Incates whether SAMPLE data should be parsed.  <em>Default: False</em></p></li>
<li><p><strong>parse_samples</strong> – If True indels will be normalized. <code class="docutils literal notranslate"><span class="pre">VCFReader</span></code> should have been
instantiated with reference. <em>Default: False</em></p></li>
<li><p><strong>check_order</strong> (<em>bool</em>) – If True VCF will be checked for sorting. <em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><strong>(chrom,it)</strong> (<em>tuple of str and iterator</em>) – <code class="docutils literal notranslate"><span class="pre">it</span></code> yields <code class="xref py py-class docutils literal notranslate"><span class="pre">variant.Genomvariant</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.VCFReader.find_vrt">
<code class="sig-name descname">find_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">2147483647</span></em>, <em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader.find_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields variant objects from a specified region</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.VCFReader.get_chroms">
<code class="sig-name descname">get_chroms</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">allow_no_index</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.VCFReader.get_chroms" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">ChromSet</span></code> corresponding to VCF. If indexed
then index is used for faster access. Alternatively if <code class="docutils literal notranslate"><span class="pre">allow_no_index</span></code>
is True the whole file is parsed to get chromosome ordering.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.vcf.BCFReader">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.vcf.</code><code class="sig-name descname">BCFReader</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bcf</span></em>, <em class="sig-param"><span class="n">index</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">reference</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader" title="Permalink to this definition">¶</a></dt>
<dd><dl class="py method">
<dt id="genomvar.vcf.BCFReader.iter_rows">
<code class="sig-name descname">iter_rows</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader.iter_rows" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields rows of variant file</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.BCFReader.find_rows">
<code class="sig-name descname">find_rows</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">rgn</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader.find_rows" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields rows of variant file</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.BCFReader.find_vrt">
<code class="sig-name descname">find_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">2147483647</span></em>, <em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader.find_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields variant objects from a specified region</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.BCFReader.get_chroms">
<code class="sig-name descname">get_chroms</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">allow_no_index</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader.get_chroms" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <code class="docutils literal notranslate"><span class="pre">ChromSet</span></code> corresponding to VCF. If indexed
then index is used for faster access. Alternatively if <code class="docutils literal notranslate"><span class="pre">allow_no_index</span></code>
is True the whole file is parsed to get chromosome ordering.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.BCFReader.get_factory">
<code class="sig-name descname">get_factory</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader.get_factory" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a factory based on normindel parameter.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.BCFReader.get_records">
<code class="sig-name descname">get_records</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader.get_records" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns parsed variant data as a dict of NumPy arrays with structured
dtype.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.BCFReader.iter_rows_by_chrom">
<code class="sig-name descname">iter_rows_by_chrom</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader.iter_rows_by_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields rows grouped by chromosome</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.BCFReader.iter_vrt">
<code class="sig-name descname">iter_vrt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader.iter_vrt" title="Permalink to this definition">¶</a></dt>
<dd><p>Yields variant objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parse_info</strong> (<em>bool</em>) – Whether INFO fields should be parsed.  <em>Default: False</em></p></li>
<li><p><strong>parse_samples</strong> (<em>bool</em>) – Whether SAMPLEs dat should be parsed.  <em>Default: False</em></p></li>
<li><p><strong>check_order</strong> (<em>bool</em>) – If True will raise exception on unsorted VCF rows. <em>Default: False</em></p></li>
<li><p><strong>normindel</strong> (<em>bool</em>) – If True insertions and deletions will be left normalized.
Requires a reference on <a class="reference internal" href="#genomvar.vcf.VCFReader" title="genomvar.vcf.VCFReader"><code class="xref py py-class docutils literal notranslate"><span class="pre">VCFReader</span></code></a> instantiation.</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><strong>vrt</strong> (<a class="reference internal" href="#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomvar.variant.GenomVariant</span></code></a>) – Variant object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf.BCFReader.iter_vrt_by_chrom">
<code class="sig-name descname">iter_vrt_by_chrom</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">parse_info</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">parse_samples</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">normindel</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">check_order</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf.BCFReader.iter_vrt_by_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates variants grouped by chromosome</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parse_info</strong> (<em>bool</em>) – Incates whether INFO fields should be parsed.  <em>Default: False</em></p></li>
<li><p><strong>parse_samples</strong> (<em>bool</em>) – Incates whether SAMPLE data should be parsed.  <em>Default: False</em></p></li>
<li><p><strong>parse_samples</strong> – If True indels will be normalized. <code class="docutils literal notranslate"><span class="pre">VCFReader</span></code> should have been
instantiated with reference. <em>Default: False</em></p></li>
<li><p><strong>check_order</strong> (<em>bool</em>) – If True VCF will be checked for sorting. <em>Default: False</em></p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><strong>(chrom,it)</strong> (<em>tuple of str and iterator</em>) – <code class="docutils literal notranslate"><span class="pre">it</span></code> yields <code class="xref py py-class docutils literal notranslate"><span class="pre">variant.Genomvariant</span></code> objects</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-vcf-utils-py">
<h2>Module vcf_utils.py<a class="headerlink" href="#module-vcf-utils-py" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-genomvar.vcf_utils"></span><dl class="py class">
<dt id="genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.vcf_utils.</code><code class="sig-name descname">VCF_INFO_OR_FORMAT_SPEC</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">NAME</span></em>, <em class="sig-param"><span class="n">NUMBER</span></em>, <em class="sig-param"><span class="n">TYPE</span></em>, <em class="sig-param"><span class="n">DESCRIPTION</span></em>, <em class="sig-param"><span class="n">SOURCE</span></em>, <em class="sig-param"><span class="n">VERSION</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC" title="Permalink to this definition">¶</a></dt>
<dd><dl class="py attribute">
<dt id="genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.DESCRIPTION">
<code class="sig-name descname">DESCRIPTION</code><a class="headerlink" href="#genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.DESCRIPTION" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 3</p>
</dd></dl>

<dl class="py attribute">
<dt id="genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.NAME">
<code class="sig-name descname">NAME</code><a class="headerlink" href="#genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.NAME" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py attribute">
<dt id="genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.NUMBER">
<code class="sig-name descname">NUMBER</code><a class="headerlink" href="#genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.NUMBER" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="py attribute">
<dt id="genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.SOURCE">
<code class="sig-name descname">SOURCE</code><a class="headerlink" href="#genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.SOURCE" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 4</p>
</dd></dl>

<dl class="py attribute">
<dt id="genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.TYPE">
<code class="sig-name descname">TYPE</code><a class="headerlink" href="#genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.TYPE" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py attribute">
<dt id="genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.VERSION">
<code class="sig-name descname">VERSION</code><a class="headerlink" href="#genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.VERSION" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 5</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.count">
<code class="sig-name descname">count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="o">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.count" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of occurrences of value.</p>
</dd></dl>

<dl class="py method">
<dt id="genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.index">
<code class="sig-name descname">index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">stop</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em>, <em class="sig-param"><span class="o">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf_utils.VCF_INFO_OR_FORMAT_SPEC.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Return first index of value.</p>
<p>Raises ValueError if the value is not present.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="genomvar.vcf_utils.VCFRow">
<em class="property">class </em><code class="sig-prename descclassname">genomvar.vcf_utils.</code><code class="sig-name descname">VCFRow</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">CHROM</span></em>, <em class="sig-param"><span class="n">POS</span></em>, <em class="sig-param"><span class="n">ID</span></em>, <em class="sig-param"><span class="n">REF</span></em>, <em class="sig-param"><span class="n">ALT</span></em>, <em class="sig-param"><span class="n">QUAL</span></em>, <em class="sig-param"><span class="n">FILTER</span></em>, <em class="sig-param"><span class="n">INFO</span></em>, <em class="sig-param"><span class="n">FORMAT</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">SAMPLES</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">rnum</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf_utils.VCFRow" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple-like object storing variant information in VCF-like form.</p>
<p>str() returns a string, formatted as a row in VCF file.</p>
</dd></dl>

<dl class="py function">
<dt id="genomvar.vcf_utils.issequence">
<code class="sig-prename descclassname">genomvar.vcf_utils.</code><code class="sig-name descname">issequence</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seq</span></em><span class="sig-paren">)</span><a class="headerlink" href="#genomvar.vcf_utils.issequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Is seq a sequence (ndarray, list or tuple)?</p>
</dd></dl>

</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Package contents</a><ul>
<li><a class="reference internal" href="#module-genomvar.variant">Module variant.py</a></li>
<li><a class="reference internal" href="#module-genomvar.varset">Module varset.py</a></li>
<li><a class="reference internal" href="#module-genomvar.vcf">Module vcf.py</a></li>
<li><a class="reference internal" href="#module-vcf-utils-py">Module vcf_utils.py</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Command_line_utility.html"
                        title="previous chapter">Comparing two VCF files from command line</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/Package_contents.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Command_line_utility.html" title="Comparing two VCF files from command line"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">genomvar  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Package contents</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, AcademGene LLC.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.4.1.
    </div>
  </body>
</html>