
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Variants &#8212; genomvar  documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Variant sets" href="Variant_sets.html" />
    <link rel="prev" title="Introduction / Quick start" href="Introduction.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Variant_sets.html" title="Variant sets"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Introduction.html" title="Introduction / Quick start"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">genomvar  documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="variants">
<h1>Variants<a class="headerlink" href="#variants" title="Permalink to this headline">¶</a></h1>
<p>Module <a class="reference internal" href="Package_contents.html#module-genomvar.variant" title="genomvar.variant"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.variant</span></code></a> contains classes representing genomic
alterations.</p>
<p>The hierarchy of the classes used in the package is the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>                  <span class="n">VariantBase</span>
                  <span class="o">/</span>      <span class="o">|</span>   \
 <span class="n">AmbigIndel</span> <span class="o">&lt;-</span><span class="n">Indel</span>      <span class="n">V</span>   <span class="o">|</span>
   <span class="o">|</span>     <span class="o">|</span>      <span class="o">/</span> \     <span class="n">MNP</span>  <span class="n">V</span>
   <span class="o">|</span> <span class="o">----+---</span> <span class="n">Del</span>  <span class="n">Ins</span>   <span class="o">|</span>  <span class="n">Haplotype</span>
   <span class="o">|</span> <span class="o">|</span>   <span class="o">|</span>          <span class="o">|</span>    <span class="n">V</span>
   <span class="n">V</span> <span class="n">V</span>   <span class="o">|</span>          <span class="n">V</span>   <span class="n">SNP</span>
<span class="n">AmbigDel</span> <span class="o">-----&gt;</span> <span class="n">AmbigIns</span>
</pre></div>
</div>
<p>All variants have <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> attributes defining a range they
act on and can be searched overlap for.</p>
<p>To test whether a variant is instance of some type
<a class="reference internal" href="Package_contents.html#genomvar.variant.VariantBase.is_instance" title="genomvar.variant.VariantBase.is_instance"><code class="xref py py-meth docutils literal notranslate"><span class="pre">is_instance()</span></code></a> method can be used.  Variant equality
can be tested using <a class="reference internal" href="Package_contents.html#genomvar.variant.VariantBase.edit_equal" title="genomvar.variant.VariantBase.edit_equal"><code class="xref py py-meth docutils literal notranslate"><span class="pre">edit_equal()</span></code></a>.</p>
<p>Objects can be instantiated directly, e.g.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vrt</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">MNP</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span><span class="mi">154678</span><span class="p">,</span><span class="s1">&#39;GT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="go">&lt;MNP chr1:154678-154680 NN/GT&gt;</span>
</pre></div>
</div>
<p>This will create an MNP which substitutes positions 154678 and 154679 on
chromosome 1 for <code class="docutils literal notranslate"><span class="pre">GT</span></code>.</p>
<p>Alternatively variants can be created using <a class="reference internal" href="Package_contents.html#genomvar.variant.VariantFactory" title="genomvar.variant.VariantFactory"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantFactory</span></code></a>
objects.  This class can work with VCF-like notation. For example,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fac</span> <span class="o">=</span> <span class="n">VariantFactory</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vrt</span> <span class="o">=</span> <span class="n">fac</span><span class="o">.</span><span class="n">from_edit</span><span class="p">(</span><span class="s1">&#39;chr15&#39;</span><span class="p">,</span><span class="mi">575</span><span class="p">,</span><span class="s1">&#39;TA&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="go">&lt;Del chr15:576-577 A/-&gt;</span>
</pre></div>
</div>
<p>Position is <strong>0-based</strong> so this creates a deletion of A in 577th
position of chromosome 15.</p>
<p>Alternatively, limited subset of HGVS notation is supported (numbering
in HGVS strings is <strong>1-based</strong> following the spec):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vrt</span> <span class="o">=</span> <span class="n">fac</span><span class="o">.</span><span class="n">from_hgvs</span><span class="p">(</span><span class="s1">&#39;chr1:g.15C&gt;A&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">vrt</span><span class="p">)</span>
<span class="go">&lt;SNP chr1:14 C/A&gt;</span>
</pre></div>
</div>
<p>Variant sets defined in <a class="reference internal" href="Variant_sets.html#module-genomvar.varset" title="genomvar.varset"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genomvar.varset</span></code></a> use class <a class="reference internal" href="Package_contents.html#genomvar.variant.GenomVariant" title="genomvar.variant.GenomVariant"><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomVariant</span></code></a>.
Objects of this class contain genomic alteration (attribute <code class="docutils literal notranslate"><span class="pre">base</span></code>)
and optionally, genotype (attribute <code class="docutils literal notranslate"><span class="pre">GT</span></code>) and other attributes
commonly found in VCF files (attribute <code class="docutils literal notranslate"><span class="pre">attrib</span></code>).  Attribute
<code class="docutils literal notranslate"><span class="pre">base</span></code> is an object of some <code class="docutils literal notranslate"><span class="pre">VariantBase</span></code> subclass (SNPs, Deletions
etc.).</p>
<div class="section" id="variant-classes">
<h2>Variant classes<a class="headerlink" href="#variant-classes" title="Permalink to this headline">¶</a></h2>
<p>Basic classes are SNP representing a single nucleotide
polimorphism and its generalization MNP – multiple nucleotide
polimorphism.</p>
<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">SNP</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">alt</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Single-nucleotide polymorphism.</p>
<p>For instantiation it requires chromosome, position and
alternative sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar</span> <span class="kn">import</span> <span class="n">variant</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variant</span><span class="o">.</span><span class="n">SNP</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span><span class="mi">154678</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">)</span>
<span class="go">SNP(&quot;chr1&quot;,154678,&quot;T&quot;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">MNP</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">alt</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Multiple-nucleotide polymorphism.  Substitute N nucleotides of the
reference for N other nucleotides.</p>
<p>For instantiation it requires chromosome,position and
alternative sequence. <code class="docutils literal notranslate"><span class="pre">end</span></code> will inferred from <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">alt</span></code>.
<code class="docutils literal notranslate"><span class="pre">ref</span></code> is also optional.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar</span> <span class="kn">import</span> <span class="n">variant</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variant</span><span class="o">.</span><span class="n">MNP</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span><span class="mi">154678</span><span class="p">,</span><span class="s1">&#39;GT&#39;</span><span class="p">)</span>
<span class="go">MNP(&quot;chr1&quot;,154678,&quot;GT&quot;)</span>
</pre></div>
</div>
</dd></dl>

<p>There are separate classes of insetion and deletion inheriting from
abstract class <code class="docutils literal notranslate"><span class="pre">Indel</span></code>.</p>
<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Ins</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">alt</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Insertion of nucleotides. For instantiation <code class="docutils literal notranslate"><span class="pre">chrom</span></code>,
<code class="docutils literal notranslate"><span class="pre">start</span></code>  and
inserted sequence (<code class="docutils literal notranslate"><span class="pre">alt</span></code>) are required.</p>
<p>Start and end denote the nucleotide after the inserted sequence,
i.e. <code class="docutils literal notranslate"><span class="pre">start</span></code> is 0-based number of a nucleotide after insertion,
<code class="docutils literal notranslate"><span class="pre">end</span></code> is <code class="docutils literal notranslate"><span class="pre">start+1</span></code> by definition.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar.variant</span> <span class="kn">import</span> <span class="n">Ins</span>
<span class="go"># Insertion of TA before position chr2:100543.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Ins</span><span class="p">(</span><span class="s1">&#39;chr2&#39;</span><span class="p">,</span><span class="mi">100543</span><span class="p">,</span><span class="s1">&#39;TA&#39;</span><span class="p">))</span>
<span class="go">&lt;Ins chr2:100543 -/TA&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Del</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alt</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Deletion of nucleotides. For instantiation <code class="docutils literal notranslate"><span class="pre">chrom</span></code>,
<code class="docutils literal notranslate"><span class="pre">start</span></code> (0-based),
and <code class="docutils literal notranslate"><span class="pre">end</span></code> (position <code class="docutils literal notranslate"><span class="pre">end</span></code> is excluded) are required.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar.variant</span> <span class="kn">import</span> <span class="n">Del</span>
<span class="go"># Deletion of 3 nucleotides starting at chr3:7843488 (0-based)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Del</span><span class="p">(</span><span class="s1">&#39;chr3&#39;</span><span class="p">,</span><span class="mi">7843488</span><span class="p">,</span><span class="mi">7843488</span><span class="o">+</span><span class="mi">3</span><span class="p">))</span>
<span class="go">&lt;Del chr3:7843488-7843491 NNN/-&gt;</span>
</pre></div>
</div>
</dd></dl>

<p>There is a special flavor of indels for cases when it can be
applied in several places resulting in the same alternate sequence,
termed to as  <code class="docutils literal notranslate"><span class="pre">ambigous</span></code> indels. They are <a class="reference internal" href="Package_contents.html#genomvar.variant.AmbigDel" title="genomvar.variant.AmbigDel"><code class="xref py py-class docutils literal notranslate"><span class="pre">AmbigDel</span></code></a>,
<a class="reference internal" href="Package_contents.html#genomvar.variant.AmbigIns" title="genomvar.variant.AmbigIns"><code class="xref py py-class docutils literal notranslate"><span class="pre">AmbigIns</span></code></a> which on top of regular deletion or insertion
attributes
contain information about a region they can be applied to. For
instantion <a class="reference internal" href="Package_contents.html#genomvar.variant.VariantFactory" title="genomvar.variant.VariantFactory"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantFactory</span></code></a> with a reference is needed.</p>
<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">AmbigIns</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">alt</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Class representing indel which position is ambigous.  Ambiguity
means the indel could be applied in any position of some region
resulting in the same alternative sequence.</p>
<p>Let the reference file <code class="docutils literal notranslate"><span class="pre">test.fasta</span></code> contain a toy sequence:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seq1</span>
<span class="n">TTTAATA</span>
</pre></div>
</div>
<p>Consider a variant extending  3 <code class="docutils literal notranslate"><span class="pre">T</span></code> s in the beginning by one
more T. It can be done in several places so the corresponding
insertion can be given as and <code class="docutils literal notranslate"><span class="pre">AmbigIns</span></code> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar</span> <span class="kn">import</span> <span class="n">Reference</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">genomvar.variant</span> <span class="kn">import</span> <span class="n">VariantFactory</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fac</span> <span class="o">=</span> <span class="n">VariantFactory</span><span class="p">(</span><span class="n">Reference</span><span class="p">(</span><span class="s1">&#39;test.fasta&#39;</span><span class="p">),</span><span class="n">normindel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span> <span class="n">fac</span><span class="o">.</span><span class="n">from_edit</span><span class="p">(</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">,</span><span class="s1">&#39;TT&#39;</span><span class="p">)</span> <span class="p">)</span>
<span class="go">&lt;AmbigIns seq1:0-4(1-2) -/T&gt;</span>
</pre></div>
</div>
<p>Positions 1 and 2 are actual start and end meaning that T
is inserted before nucleotide located 1-2.
Positions 0-4 indicate that start and end can be extended to
these values resulting in the same alteration.</p>
</dd></dl>

<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">AmbigDel</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alt</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Class representing del which position is ambigous.  Ambiguity means
the same number of positions could be deleted in some range
resulting in the same aternative sequence.</p>
<p>Let the reference file <code class="docutils literal notranslate"><span class="pre">test.fasta</span></code> contain a toy sequence:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">seq1</span>
<span class="n">TCTTTTTGACTGG</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fac</span> <span class="o">=</span> <span class="n">VariantFactory</span><span class="p">(</span><span class="n">Reference</span><span class="p">(</span><span class="s1">&#39;test.fasta&#39;</span><span class="p">),</span><span class="n">normindel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">&gt;&gt;&gt;</span>
<span class="go">print( fac.from_edit(&#39;seq1&#39;,1,&#39;CTTTTTGAC&#39;,&#39;C&#39;) ) &lt;AmbigDel</span>
<span class="go">seq1:1-11(2-10) TTTTTGAC/-&gt;</span>
</pre></div>
</div>
<p>Deletion of TTTTTGAC starts at 2 and ends on 9th nucleotide
(including 9th resulting in range 2-10).  1-11 denote that start and
end can be extended to these values resulting in the same
alteration.</p>
</dd></dl>

<p>There is a separate class <code class="docutils literal notranslate"><span class="pre">Haplotype</span></code> which can hold any combination
of objects of variant classes above.</p>
<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">genomvar.variant.</code><code class="sig-name descname">Haplotype</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chrom</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em>, <em class="sig-param"><span class="n">variants</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>An object representing genome variants on the
same chromosome (or contig).</p>
<p>Can be instantiated from a list of GenomVariant objects using
<a class="reference internal" href="Package_contents.html#genomvar.variant.Haplotype.from_variants" title="genomvar.variant.Haplotype.from_variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Haplotype.from_variants()</span></code></a> class method.</p>
</dd></dl>

<p>Additionally module contains two technically driven types:
Null (no variant at all, reference),
Asterisk (for * in the ALT field of VCF files).</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Variants</a><ul>
<li><a class="reference internal" href="#variant-classes">Variant classes</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Introduction.html"
                        title="previous chapter">Introduction / Quick start</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Variant_sets.html"
                        title="next chapter">Variant sets</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/Variants.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Variant_sets.html" title="Variant sets"
             >next</a> |</li>
        <li class="right" >
          <a href="Introduction.html" title="Introduction / Quick start"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">genomvar  documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, AcademGene LLC.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.0.0.
    </div>
  </body>
</html>